hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java

```diff
@@ -53,6 +54,41 @@ public CustomGasCharging(@NonNull final GasCalculator gasCalculator) {
         this.gasCalculator = gasCalculator;
     }
 
+    /**
+     * If the actual gas used by a transaction merits a refund, then do the refund using
+     * relayer-aware logic. The relayer gets refund priority, so any charged allowance
+     * is first refunded, with the sender only getting a refund if there is any left over.
+     *
+     * @param unusedGas the actual gas used by the transaction
+     * @param allowanceUsed the amount of the relayer's gas allowance used
+     * @param sender the sender account
+     * @param relayer the relayer account, if present
+     * @param context the context of the transaction, including the network gas price
+     * @param worldUpdater the world updater for the transaction
+     */
+    public void maybeRefundGiven(
+            final long unusedGas,
+            final long allowanceUsed,
+            @NonNull final HederaEvmAccount sender,
+            @Nullable final HederaEvmAccount relayer,
+            @NonNull final HederaEvmContext context,
+            @NonNull final HederaWorldUpdater worldUpdater) {
+        requireAllNonNull(sender, context, worldUpdater);
+        if (context.isNoopGasContext() || unusedGas == 0) {
+            return;
+        }
+        final var refund = unusedGas * context.gasPrice();
+        if (allowanceUsed > 0) {
+            requireNonNull(relayer);
```

Comment:
I don't get it exactly. The argument `relayer` is `@Nullable` but all of the sudden here it is `requireNonNull`.  What changed - that `allowanceUsed > 0`?  That could be because at the caller `TransactionProcessor.computeInvolvedParties` the thing it is passing to `allowanceUsed` here is called `relayerAllowancedUsed` there.  But here, without that "relayer" on the `allowancedUsed` argument name, seems unsupported.

BTW, instead of using an if/else at L81 & L87 you could directly implement the method comment by making `unusedGas` _not_ `final` - then first handle the relayer's refund, decrement `unusedGas` accordingly, then handle the remainder.  Untangling it that way may make it easier to read.

Reply:
> What changed - that allowanceUsed > 0? 

Right, it would be an invariant failure to return a non-zero `allowanceUsed`, but without a `relayer` account to have _paid_ that allowance. Should never happen for sure.

> BTW, instead of using an if/else at L81 & L87 you could directly implement the method comment by making unusedGas not final - then first handle the relayer's refund, decrement unusedGas accordingly...

Not sure I follow this one. 🤔